lib/main_common.pm

```diff
@@ -2401,116 +2401,186 @@ sub load_virtualization_tests {
-        loadtest 'virtualization/universal/upgrade_guests';    # Upgrade all guests
-        loadtest 'virtualization/universal/patch_guests';    # Apply patches to all compatible guests
-        loadtest 'virtualization/universal/patch_and_reboot';    # Apply updates and reboot
-
+    # QAM VM installation
+    unless (get_var('DISABLE_VM_INSTALL')) {
         loadtest "virt_autotest/login_console";
-    }
-
-    loadtest "virtualization/universal/list_guests" unless ($windows);    # List all guests and ensure they are running
-
-    if (check_var('VIRT_PART', 'install')) {
-        loadtest "virtualization/universal/kernel";    # Virtualization kernel functions
-    }
-
-    if (check_var('VIRT_PART', 'virtmanager')) {
-        loadtest 'virtualization/universal/virtmanager_init';    # Connect to the Xen hypervisor using virt-manager
-        loadtest 'virtualization/universal/virtmanager_offon';    # Turn all VMs off and then on again
-
-        if (is_sle('12-SP3+')) {
-            loadtest 'virtualization/universal/virtmanager_add_devices';    # Add some aditional HV to all VMs
-            loadtest 'virtualization/universal/virtmanager_rm_devices';    # Remove the aditional HV from all VMs
+        if (check_var('INSTALL_GUEST', 1)) {
+            # Prepare and wait for guests
+            loadtest "virtualization/universal/prepare_guests";
+            loadtest "virtualization/universal/waitfor_guests";
+        }
+        if (check_var('PATCH_WITH_ZYPPER', 1)) {
+            # Patch system and reboot
+            loadtest "virtualization/universal/patch_and_reboot";
+            if (my $update_package = get_var('UPDATE_PACKAGE')) {
+                if ($update_package eq 'kernel-default') {
+                    # If update package is 'kernel-default', list and patch guests
+                    loadtest "virt_autotest/login_console";
+                    loadtest "virtualization/universal/list_guests";
+                    loadtest "virtualization/universal/patch_guests";
+                } elsif ($update_package eq 'xen' || $update_package eq 'qemu') {
+                    # If update package is 'xen' or 'qemu', list guests
+                    loadtest "virt_autotest/login_console";
+                    loadtest "virtualization/universal/list_guests";
+                }
+            }
+        }
+        # Run kernel tests and finish
+        loadtest "virtualization/universal/kernel";
+        loadtest "virtualization/universal/finish";
+    }
+
+    # Only execute the following tests if UPDATE_PACKAGE is not 'kernel-default'
+    unless (check_var('UPDATE_PACKAGE', 'kernel-default')) {
+        # QAM VIRT save_and_restore test
+        unless (get_var('DISABLE_SAVE_AND_RESTORE')) {
+            loadtest "console/login";
+            loadtest "virt_autotest/cleanup_libvirtd_log";
+            loadtest "virtualization/universal/save_and_restore";
+            loadtest "virtualization/universal/finish";
+        }
```

Comment:
It seems repeating for below pattern, which does not follow `DRY` (do not repeat yourself) rule. Would you please wrap it into some function like `check_and_load_mu_virt_features`, and reuse it by giving  input of  file names.


```
        unless (get_var('DISABLE_xx')) {
            loadtest "console/login";
            loadtest "virt_autotest/cleanup_libvirtd_log";
            loadtest ...
            loadtest "virtualization/universal/finish";
        }
```

Reply:
I will not think about loop temporarily, it is difficult to deal with it.
Sometime still need check xen or kvm.
After code is stable, will think about such plan.

Reply:
I think you can do it with sth like below . For DOM_METRICS and IRQBALANCE, the xen checking can be moved out of switch check. Then all can reuse this, except that the 2 INSTALL related ones  are still too complex to use it.


```
sub check_and_load_mu_virt_features {
    my ($switch, $test_modules) = @_;
    die "You can't have empty switch or test_modules" if (!switch or !$test_modules);
    
    unless (get_var("$switch")) {
            loadtest "console/login";
            loadtest "virt_autotest/cleanup_libvirtd_log";
            loadtest "$module" foreach my $module (split(',', $test_modules));
            loadtest "virtualization/universal/finish";
     }
}


# Then call it for feature tests
check_and_load_mu_virt_features('DISABLE_SAVE_AND_RESTORE', 'virtualization/universal/save_and_restore');

check_and_load_mu_virt_features('DISABLE_GUEST_MANAGEMENT ', 'virtualization/universal/guest_management');

check_and_load_mu_virt_features('DISABLE_FINAL_TEST',         'virtualization/universal/ssh_final,virtualization/universal/virtmanager_final,virtualization/universal/smoketest,virtualization/universal/stresstest,console/perf')

...
```

Reply:
updated, please check, a little difference.

Reply:
@RoyCai7 Hi Roy, I checked the latest code. I am afraid it is not decreasing the repeating lines by putting test modules into some variable, and it makes the logic less direct.

Any reason that you do not pick the way in my above previous comment? Data structure for test_modules is not well readable? If yes, this part can definitely be improved, it can be array or sth else.

Reply:
Updated code by your method, please check.